Subject: [PATCH] fat: fix buffer overflow in vfat_create_shortname()
From: Fedor <fedork@ubuntu.(none)>
Date: Fri, 2 Apr 2010 12:08:21 -0400

backport from 2.6
(original kernel.org commit 30d1872d9eb3663b4cf7bdebcbf5cd465674cced)
---
 fs/vfat/namei.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/fs/vfat/namei.c b/fs/vfat/namei.c
index 4437d5d..0cc0142 100644
--- a/fs/vfat/namei.c
+++ b/fs/vfat/namei.c
@@ -520,7 +520,7 @@ static int vfat_create_shortname(struct inode *dir, struct nls_table *nls,
 				 char *name_res, unsigned char *lcase)
 {
 	wchar_t *ip, *ext_start, *end, *name_start;
-	unsigned char base[9], ext[4], buf[8], *p;
+	unsigned char base[9], ext[4], buf[5], *p;
 	unsigned char charbuf[NLS_MAX_CHARSET_SIZE];
 	int chl, chi;
 	int sz = 0, extlen, baselen, i, numtail_baselen, numtail2_baselen;
@@ -677,7 +677,7 @@ static int vfat_create_shortname(struct inode *dir, struct nls_table *nls,
 			return 0;
 	}
 
-	i = jiffies & 0xffff;
+	i = jiffies;
 	sz = (jiffies >> 16) & 0x7;
 	if (baselen>2) {
 		baselen = numtail2_baselen;
@@ -686,7 +686,7 @@ static int vfat_create_shortname(struct inode *dir, struct nls_table *nls,
 	name_res[baselen+4] = '~';
 	name_res[baselen+5] = '1' + sz;
 	while (1) {
-		sprintf(buf, "%04X", i);
+		snprintf(buf, sizeof(buf), "%04X", i & 0xffff);
 		memcpy(&name_res[baselen], buf, 4);
 		if (vfat_find_form(dir, name_res) < 0)
 			break;
-- 
1.6.5.GIT

